### PR TITLE
fix(nm): improves external soft links hoisting

### PR DESCRIPTION
**Pull request by larixer**
**What's the problem this PR addresses?**

When all the dependencies were hoisted from nested portals the portals were not hoisted

Fixes: #4240

**How did you fix it?**

Applies additional hoisting rounds when external soft links run out of unhoisted dependencies

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-nm/sources/hoist.ts [hunk 1]
```diff
@@ -396,7 +396,12 @@ const hoistTo = (tree: HoisterWorkTree, rootNodePath: Array<HoisterWorkTree>, ro
 
   const hoistIdentMap = getHoistIdentMap(rootNode, preferenceMap);
 
-  const usedDependencies = tree == rootNode ? new Map() : (options.fastLookupPossible ? getZeroRoundUsedDependencies(rootNodePath) :  getUsedDependencies(rootNodePath));
+  const usedDependencies = tree == rootNode ? new Map() :
+    (options.fastLookupPossible
+      ? getZeroRoundUsedDependencies(rootNodePath)
+      :  getUsedDependencies(rootNodePath)
```
Comment (trivikr) [inline packages/yarnpkg-nm/sources/hoist.ts:402]: nit: remove extra space
```suggestion
      : getUsedDependencies(rootNodePath)
```

File: packages/yarnpkg-nm/sources/hoist.ts [hunk 2]
```diff
@@ -438,6 +443,17 @@ const hoistTo = (tree: HoisterWorkTree, rootNodePath: Array<HoisterWorkTree>, ro
...
 
+const hasUnhoistedDependencies = (node: HoisterWorkTree): boolean => {
+  let result = false;
+  for (const [subName, subDependency] of node.dependencies) {
+    if (!node.peerNames.has(subName) && subDependency.ident !== node.ident) {
+      result = true;
+      break;
+    }
+  }
+  return result;
```
Comment (trivikr) [inline packages/yarnpkg-nm/sources/hoist.ts:454]: Skip using variable `result`?
```suggestion
  for (const [subName, subDependency] of node.dependencies) {
    if (!node.peerNames.has(subName) && subDependency.ident !== node.ident) {
      return true
    }
  }
  return false;
```

**Review by merceyz — APPROVED** [review]

File: packages/yarnpkg-nm/sources/hoist.ts [hunk 3]
```diff
@@ -668,6 +684,9 @@ const hoistGraph = (tree: HoisterWorkTree, rootNodePath: Array<HoisterWorkTree>,
       }
     }
 
+    if (parentNode.dependencyKind === HoisterDependencyKind.EXTERNAL_SOFT_LINK && wereNodesHoisted)
```
Comment (pilaoda) [inline packages/yarnpkg-nm/sources/hoist.ts:687]: Why only EXTERNAL_SOFT_LINK need anotherRound? I meet a bug that parentNode is a sub workspace and could not hoist its dependencies.

Comment (merceyz) [reply packages/yarnpkg-nm/sources/hoist.ts:687]: Please open an issue with a reproduction.

Comment (pilaoda) [reply packages/yarnpkg-nm/sources/hoist.ts:687]: #5332 here is the issue with a reproduction repo.

**Pull request merged**